librz/bin/p/bin_java.c

```diff
@@ -1,227 +1,259 @@
-	ret->machine = strdup("jvm");
-	ret->arch = strdup("java");
-	ret->bits = 32;
-	ret->big_endian = 0;
-	ret->dbg_info = 4 | 8; /* LineNums | Syms */
-	return ret;
+	tmp = rz_bin_java_class_fields_as_symbols(jclass);
+	rz_list_join(list, tmp);
+	rz_list_free(tmp);
+
+	tmp = rz_bin_java_class_const_pool_as_symbols(jclass);
+	rz_list_join(list, tmp);
+	rz_list_free(tmp);
+	return list;
 }
 
-static bool check_buffer(RzBuffer *b) {
-	if (rz_buf_size(b) > 32) {
-		ut8 buf[4];
-		rz_buf_read_at(b, 0, buf, sizeof(buf));
-		if (!memcmp(buf, "\xca\xfe\xba\xbe", 4)) {
-			int off = rz_buf_read_be32_at(b, 4 * sizeof(int));
-			int version = rz_buf_read_be16_at(b, 6);
-			if (off > 0 && version < 1024) {
-				return true;
-			}
-		}
+static RzList *fields(RzBinFile *bf) {
+	RzBinJavaClass *jclass = rz_bin_file_get_java_class(bf);
+	if (!jclass) {
+		return NULL;
 	}
-	return false;
+
+	return rz_bin_java_class_fields_as_binfields(jclass);
 }
 
-static int retdemangle(const char *str) {
-	return RZ_BIN_NM_JAVA;
+static RzList *libs(RzBinFile *bf) {
+	RzBinJavaClass *jclass = rz_bin_file_get_java_class(bf);
+	if (!jclass) {
+		return NULL;
+	}
+
+	return rz_bin_java_class_as_libraries(jclass);
 }
 
 static RzBinAddr *binsym(RzBinFile *bf, int sym) {
-	return rz_bin_java_get_entrypoint(bf->o->bin_obj, sym);
-}
+	RzBinJavaClass *jclass = rz_bin_file_get_java_class(bf);
+	if (!jclass) {
+		return NULL;
+	}
 
-static RZ_BORROW RzList *lines(RzBinFile *bf) {
-	return NULL;
-#if 0
-	char *file = bf->file? strdup (bf->file): strdup ("");
-	RzList *list = rz_list_newf (free);
-	// XXX the owner of this list should be the plugin, so we are leaking here
-	file = rz_str_replace (file, ".class", ".java", 0);
-	/*
-	   int i;
-	   RzBinJavaObj *b = bf->o->bin_obj;
-	   for (i=0; i<b->lines.count; i++) {
-	        RzBinDwarfRow *row = RZ_NEW0(RzBinDwarfRow);
-	        rz_bin_dwarf_line_new (row, b->lines.addr[i], file, b->lines.line[i]);
-	        rz_list_append (list, row);
-	   }*/
-	free (file);
-	return list;
-#endif
+	return rz_bin_java_class_resolve_symbol(jclass, sym);
 }
 
-static RzList *sections(RzBinFile *bf) {
-	return rz_bin_java_get_sections(bf->o->bin_obj);
+static RzList *entrypoints(RzBinFile *bf) {
+	RzBinJavaClass *jclass = rz_bin_file_get_java_class(bf);
+	if (!jclass) {
+		return NULL;
+	}
+
+	return rz_bin_java_class_entrypoints(jclass);
 }
 
-static RzList *imports(RzBinFile *bf) {
-	return rz_bin_java_get_imports(bf->o->bin_obj);
+static RzList *strings(RzBinFile *bf) {
+	RzBinJavaClass *jclass = rz_bin_file_get_java_class(bf);
+	if (!jclass) {
+		return NULL;
+	}
+
+	return rz_bin_java_class_strings(jclass);
 }
 
-static RzList *fields(RzBinFile *bf) {
-	return NULL; // rz_bin_java_get_fields (bf->o->bin_obj);
+static int demangle_type(const char *str) {
+	return RZ_BIN_NM_JAVA;
 }
 
-static RzList *libs(RzBinFile *bf) {
-	return rz_bin_java_get_lib_names(bf->o->bin_obj);
+static char *enrich_asm(RzBinFile *bf, const char *asm_str, int asm_len) {
```

Comment:
what is exactly the purpose of this new plugin method? Do we really need it? If it's going to modify the asm, isn't a RzParsePlugin better?

Reply:
essentially, as explained on MM, on java you have a table with indices (from 0 to N) with all the references (like strings, definitions, const values, etc..) and when you load one of these, you have just the number of the index, not a file offset/address. so in order to enrich the asm to have `ldc constant_pool.11` you will have `ldc 1.5875f` for example.
This is just the cleanest way to fix this issue on asm and to avoid to have rzasm depending on rzbin

Reply:
:( :( I hate these things... discussing on Mattermost right now, but I guess for now it's fine :( :(